File: app/src/main/java/danb/speedrunbrowser/MessagingService.java
package danb.speedrunbrowser;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;

import java.io.IOException;
import java.net.URL;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import danb.speedrunbrowser.api.PushNotificationData;
import danb.speedrunbrowser.api.objects.User;
import danb.speedrunbrowser.utils.Constants;
import danb.speedrunbrowser.utils.Util;
import okhttp3.Request;
import okhttp3.Response;

public class MessagingService extends FirebaseMessagingService {
    private static final String TAG = MessagingService.class.getSimpleName();

    public MessagingService() {
    }

    @Override
    public void onMessageReceived(RemoteMessage remoteMessage) {
        super.onMessageReceived(remoteMessage);

        Log.d(TAG, "From: " + remoteMessage.getFrom());

        // Check if message contains a notification payload.
        if (remoteMessage.getData() != null && remoteMessage.getFrom() != null) {
            Map<String, String> rawData = remoteMessage.getData();
            Log.d(TAG, "Message Data Body: " + rawData);

            PushNotificationData data = new PushNotificationData(rawData);

            if(remoteMessage.getFrom().startsWith("/topics/" + getBuildVariant() + "_player")) {
                makePlayerNotification(data);
            }
            else if(remoteMessage.getFrom().startsWith("/topics/" + getBuildVariant() + "_game")) {
                makeGameNotification(data);
            }

        }
    }

    private void makePlayerNotification(PushNotificationData data) {

        List<User> players = data.getNew_run().getRun().getPlayers();

        if(players.isEmpty()) {
            Log.w(TAG, "Received notification of run with no players! Cancelling notification.");
            return;
        }

        // download feature image from the internet
        // TODO: Utilize in-app cache in the future
        Bitmap featureImg = null;
        try {

            URL playerImgUrl = new URL(String.format(Constants.AVATAR_IMG_LOCATION, players.get(0).getNames().get("international")));

            Log.d(TAG, "Download player img from:" + playerImgUrl);

            Request req = new Request.Builder()
                    .url(playerImgUrl)
                    .build();

            Response res = Objects.requireNonNull(Util.getHTTPClient()).newCall(req).execute();

            if(!res.isSuccessful())
                throw new IOException();


            featureImg = BitmapFactory.decodeStream(res.body().byteStream());
        } catch(IOException e) {
            Log.e(TAG, "Could not download player feature img:", e);
        }

        String title = getString(R.string.notify_title_player_record, User.Companion.printPlayerNames(players), data.getGame().getResolvedName());

        String categoryAndLevelName = data.getCategory().getName();
        if(data.getLevel() != null)
            categoryAndLevelName += " - " + data.getLevel().getName();

        String msg = getString(R.string.notify_msg_player_record, data.getNew_run().getPlaceName() + " place",
                data.getGame().getResolvedName(), categoryAndLevelName,
                data.getOld_run() != null ? data.getOld_run().getRun().getTimes().getTime() : "", data.getNew_run().getRun().getTimes().getTime());

        Intent intent = new Intent(this, ItemDetailActivity.class);
        intent.putExtra(ItemDetailActivity.EXTRA_ITEM_TYPE, ItemListFragment.ItemType.PLAYERS);
        intent.putExtra(PlayerDetailFragment.ARG_PLAYER_ID, players.get(0).getId());

        Util.postNotification(this, intent, players.get(0).getId(), title, msg, featureImg);
    }

    private void makeGameNotification(PushNotificationData data) {
        List<User> players = data.getNew_run().getRun().getPlayers();

        if(players.isEmpty()) {
            Log.w(TAG, "Received notification of run with no players! Cancelling notification.");
            return;
        }

        // download feature image from the internet
        // TODO: Utilize in-app cache in the future
        Bitmap featureImg = null;
        try {

            URL gameCoverUrl = data.getGame().getAssets().getCoverLarge().getUri();

            Log.d(TAG, "Download player img from:" + gameCoverUrl);

            Request req = new Request.Builder()
                    .url(gameCoverUrl)
                    .build();

            Response res = Objects.requireNonNull(Util.getHTTPClient()).newCall(req).execute();

            if(!res.isSuccessful())
                throw new IOException();


            featureImg = BitmapFactory.decodeStream(res.body().byteStream());
        } catch(IOException e) {
            Log.e(TAG, "Could not download player feature img:", e);
        }

        String title = getString(R.string.notify_title_game_record, data.getGame().getResolvedName(), User.Companion.printPlayerNames(players));

        String categoryAndLevelName = data.getCategory().getName();

        if(data.getLevel() != null)
            categoryAndLevelName += " - " + data.getLevel().getName();

        String msg = getString(R.string.notify_msg_game_record, data.getNew_run().getPlaceName() + " place",
                data.getGame().getResolvedName(), categoryAndLevelName,
                data.getOld_run() != null ? data.getOld_run().getRun().getTimes().getTime() : "", data.getNew_run().getRun().getTimes().getTime());

        Intent intent = new Intent(this, ItemDetailActivity.class);
        intent.putExtra(ItemDetailActivity.EXTRA_ITEM_TYPE, ItemListFragment.ItemType.GAMES);
        intent.putExtra(GameDetailFragment.ARG_GAME_ID, data.getGame().getId());

        Util.postNotification(this, intent, data.getGame().getId(), title, msg, featureImg);
    }

    private String getBuildVariant() {
        return BuildConfig.DEBUG ? "debug" : "release";
    }
}

File: app/src/main/java/danb/speedrunbrowser/MessagingService.kt
package danb.speedrunbrowser

import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log

import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage

import java.io.IOException
import java.net.URL
import java.util.Objects

import danb.speedrunbrowser.api.PushNotificationData
import danb.speedrunbrowser.api.objects.User
import danb.speedrunbrowser.utils.Constants
import danb.speedrunbrowser.utils.ItemType
import danb.speedrunbrowser.utils.Util
import okhttp3.OkHttpClient
import okhttp3.Request

class MessagingService : FirebaseMessagingService() {

    private val buildVariant: String
        get() = if (BuildConfig.DEBUG) "debug" else "release"

    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        super.onMessageReceived(remoteMessage)

        Log.d(TAG, "From: " + remoteMessage.from!!)

        // Check if message contains a notification payload.
        if (remoteMessage.from != null) {
            val rawData = remoteMessage.data
            Log.d(TAG, "Message Data Body: $rawData")

            val data = PushNotificationData(rawData)

            if (remoteMessage.from!!.startsWith("/topics/" + buildVariant + "_player")) {
                makePlayerNotification(data)
            } else if (remoteMessage.from!!.startsWith("/topics/" + buildVariant + "_game")) {
                makeGameNotification(data)
            }

        }
    }

    private fun makePlayerNotification(data: PushNotificationData) {

        val players = data.newRun.run.players

        if (players!!.isEmpty()) {
            Log.w(TAG, "Received notification of run with no players! Cancelling notification.")
            return
        }

        // download feature image from the internet
        // TODO: Utilize in-app cache in the future
        var featureImg: Bitmap? = null
        try {

            val playerImgUrl = URL(String.format(Constants.AVATAR_IMG_LOCATION, players[0].names!!["international"]))

            Log.d(TAG, "Download player img from:$playerImgUrl")

            val req = Request.Builder()
                    .url(playerImgUrl)
                    .build()

            val res = Objects.requireNonNull<OkHttpClient>(Util.getHTTPClient(this)).newCall(req).execute()

            if (!res.isSuccessful)
                throw IOException()


            featureImg = BitmapFactory.decodeStream(res.body!!.byteStream())
        } catch (e: IOException) {
            Log.e(TAG, "Could not download player feature img:", e)
        }

        val title = getString(R.string.notify_title_player_record, User.printPlayerNames(players), data.game.resolvedName)

        var categoryAndLevelName = data.category.name
        if (data.level != null)
            categoryAndLevelName += " - " + data.level.name


        val msg = getString(R.string.notify_msg_player_record, data.newRun.placeName + " place",
                categoryAndLevelName, data.newRun.run.times!!.time)

        val intent = Intent(this, SpeedrunBrowserActivity::class.java)
        intent.putExtra(SpeedrunBrowserActivity.EXTRA_FRAGMENT_CLASSPATH, PlayerDetailFragment::class.java.canonicalName)
        intent.putExtra(PlayerDetailFragment.ARG_PLAYER_ID, players[0].id)

        Util.postNotification(this, intent, players[0].id, title, msg, featureImg)
    }

    private fun makeGameNotification(data: PushNotificationData) {
        val players = data.newRun.run.players

        if (players!!.isEmpty()) {
            Log.w(TAG, "Received notification of run with no players! Cancelling notification.")
            return
        }

        // download feature image from the internet
        // TODO: Utilize in-app cache in the future
        var featureImg: Bitmap? = null
        try {

            val gameCoverUrl = data.game.assets.coverLarge!!.uri

            Log.d(TAG, "Download player img from:$gameCoverUrl")

            val req = Request.Builder()
                    .url(gameCoverUrl)
                    .build()

            val res = Objects.requireNonNull<OkHttpClient>(Util.getHTTPClient(this)).newCall(req).execute()

            if (!res.isSuccessful)
                throw IOException()


            featureImg = BitmapFactory.decodeStream(res.body!!.byteStream())
        } catch (e: IOException) {
            Log.e(TAG, "Could not download player feature img:", e)
        }

        val title = getString(R.string.notify_title_game_record, data.game.resolvedName, User.printPlayerNames(players))

        var categoryAndLevelName = data.category.name

        if (data.level != null)
            categoryAndLevelName += " - " + data.level.name

        val msg = getString(R.string.notify_msg_game_record, data.newRun.placeName + " place",
                categoryAndLevelName, data.newRun.run.times!!.time)

        val intent = Intent(this, SpeedrunBrowserActivity::class.java)
        intent.putExtra(SpeedrunBrowserActivity.EXTRA_FRAGMENT_CLASSPATH, GameDetailFragment::class.java.canonicalName)
        intent.putExtra(GameDetailFragment.ARG_GAME_ID, data.game.id)

        Util.postNotification(this, intent, data.game.id, title, msg, featureImg)
    }

    companion object {
        private val TAG = MessagingService::class.java.simpleName
    }
}
